improver/generate_ancillaries/generate_derived_solar_fields.py

```diff
@@ -177,6 +184,109 @@ def _irradiance_times(
+        Approximation Formula", Applied Optics, vol. 28, p. 4735-4738, 1989.
+        [2] M. Reno, C. Hansen, and J. Stein, "Global Horizontal Irradiance
+        Clear Sky Models: Implementation and Analysis", Sandia National
+        Laboratories, SAND2012-2389, 2012.
+        """
+        # For angles with zenith > 90, the air_mass is ill-defined. Here we
+        #  acknowledge that these values will result in invalid values in the
+        #  power calculation.
+        zenith_above_horizon = np.where(np.abs(zenith) >= 90, np.nan, zenith)
+        with warnings.catch_warnings():
+            warnings.filterwarnings("ignore", "invalid value encountered in power")
+
+            air_mass = 1.0 / (
+                np.cos(np.radians(zenith_above_horizon))
+                + 0.50572 * (96.07995 - zenith_above_horizon) ** (-1.6364)
+            )
+        # Remove nans_associated with zenith > 90.0 degrees
+        air_mass = np.nan_to_num(air_mass)
+
+        return air_mass
+
+    def _calc_clearsky_ineichen(
+        self,
+        zenith_angle: ndarray,
+        day_of_year: int,
+        surface_altitude: Union[ndarray, float],
+        linke_turbidity: Union[ndarray, float],
+    ) -> ndarray:
+        """Calculate the clearsky global horizontal irradiance using the Perez
+        & Ineichen (2002) [1, 2] formulation. Note that the formulation here
+        neglects the Perez enhancement that can be found in the literature; see
+        PvLib issue (https://github.com/pvlib/pvlib-python/issues/435) for details.
+
+        Note: this method produces values that exceed the incoming extra-terrestrial
+        irradiance for large altitude values (> 5000m). Here we limit the the value to
+        not exceed the incoming irradiance. Caution should be used for irradiance values
+        at large altitudes as they are likely to over-estimate.
+
+        Args:
+            zenith_angle:
+                zenith_angle angle in degrees.
+            day_of_year:
+                Day of the year.
+            surface_altitude:
+                Grid box elevation.
+            linke_turbidity:
+                Linke_turbidity value is a dimensionless value that characterises the
+                atmospheres ability to scatter incoming radiation relative to a dry
+                atmosphere.
+
+        Returns:
+            Clearsky global horizontal irradiance values.
+
+        References:
+            [1] INEICHEN, Pierre, PEREZ, R. "A new airmass independent formulation
+            for the Linke turbidity coefficient", Solar Energy, vol. 73, p. 151-157,
+            2002.
+            [2] M. Reno, C. Hansen, and J. Stein, "Global Horizontal Irradiance
+            Clear Sky Models: Implementation and Analysis", Sandia National
+            Laboratories, SAND2012-2389, 2012.
+        """
+        # Day of year as an angular quantity.
+        theta0 = 2 * np.pi * day_of_year / 365.0
```

Comment:
The values for `day_of_year` are zero-indexed, so all values are expected to be between 0 and 365 (with 365 -> 366th day of the year). This will mean `theta0` values will be between `0.0` and `2 * np.pi`, with the RH bound only occurring in leap-years; this will mean that on leap years, the value for theta0 will effectively be the same for two consecutive days (when trig functions are applied). So the answers is yes, leap years will effect this.

However, I'm reluctant to change this for the following reason:
* The calculation of irradiance uses zenith which is evaluated using the [calc_solar_elevation](https://github.com/metoppv/improver/blob/6213f36a5086e081a199030d9bea39e3b9251715/improver/utilities/solar.py#L121) in solor.py. This function in turn calls `calc_solar_hour_angle` which evaluates `theta0` as above. Making the suggested change will require updating both instances. As the solar routines are used by other CLIs, I'd want to be sure that this change won't have any unwanted side-effects to other IMPROVER functionality. It would also require passing in additional arguments to `calc_solar_hour_angle` to allow for the determination of the number of days in a year to use.

The impact of using 365 in all cases rather than the number of days in a given year will be small and for our usage I wouldn't expect there to be any noticeable effect. So at this stage I'm inclined to say leave it.

Reply:
Fair enough, thanks for the detailed answer!